File: app/src/main/java/com/example/companionek/diary_activity.java
package com.example.companionek;

import android.app.Activity;
import android.os.Bundle;


import android.view.View;
import android.widget.EditText;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.ImageView;

public class diary_activity extends Activity {

	
	private View _bg__diary;
	private View rectangle_6;
	private ScrollView diary;
	private View _bg__x_ek1;
	private ImageView vector;
	private View _bg__frame_1_ek1;
	private TextView wednesday__march_27___2024;
	private TextView feeling_stuck;
	private TextView what_s_making_you_feel_stuck__;
	private TextView feeling_stuck_is_a_common_experience_we_all_encounter_at_times__it_s_that_frustrating_sensation_when_we_find_ourselves_unable_to_make_progress_or_find_a_solution_to_a_problem__it_can_manifest_in_various_aspects_of_life__such_as_work__relationships__or_personal_goals__being_stuck_often_arises_from_a_lack_of_clarity__motivation_;
	private View _bg__frame_3_ek1;
	private View _bg__frame_2_ek1;
	private View _bg__microphone_ek1;
	private ImageView vector_ek1;
	private View _bg__spotifylogo_ek1;
	private ImageView vector_ek2;
	private View _bg__camera_ek1;
	private ImageView vector_ek3;
	private View _bg__trash_ek1;
	private ImageView vector_ek4;
	private View _bg__frame_28_ek1;
	private View _bg__chevron_left_ek1;
	private ImageView back_vector;
	private View _bg__chevron_left_ek3;
	private ImageView vector_ek6;
	private View _bg__addnew_ek1;
	private View ellipse_6;
	private View _bg__plus_ek1;
	private ImageView vector_ek7;
	private ImageView vector_ek8;

	@Override
	public void onCreate(Bundle savedInstanceState) {

		super.onCreate(savedInstanceState);
		setContentView(R.layout.diary);


		wednesday__march_27___2024 = (TextView) findViewById(R.id.wednesday__march_27___2024);
		back_vector = (ImageView) findViewById(R.id.back_vector);
	
		
		//custom code goes here
		back_vector.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View view) {
				finish();
			}
		});
	
	}
}
	
	

File: app/src/main/java/com/example/companionek/diary_activity.kt
package com.example.companionek

import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.Toast
import com.example.companionek.data.Note
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore

class diary_activity : Activity() {


    private var back_vector: ImageView? = null
    private lateinit var save_button:ImageView
    private lateinit var delete_button:ImageView
    private lateinit var titleEditText:EditText
    private lateinit var noteContentEditText:EditText
    private lateinit var title:String
    private lateinit var content:String
    private val firestore = FirebaseFirestore.getInstance()
    private var noteId:String? = null

    private var  userId = FirebaseAuth.getInstance().currentUser?.uid

    @SuppressLint("MissingInflatedId")
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.diary)
        back_vector = findViewById<View>(R.id.backButton) as ImageView
        save_button=findViewById(R.id.saveButton)
        delete_button=findViewById(R.id.ic_delete_button)
        titleEditText=findViewById(R.id.titleEditText)
        noteContentEditText=findViewById(R.id.noteEditText)
        noteId = intent.getStringExtra("NOTE_ID")?:null
        Log.d("noteId: ","$noteId")
        if (noteId != null) {
            loadNoteContent()
            delete_button.setOnClickListener {
                AlertDialog.Builder(this)
                    .setTitle("Confirm Deletion")
                    .setMessage("Are you sure you want to delete this note?")
                    .setPositiveButton("Yes") { _, _ ->
                        // Delete note if confirmed
                        val noteRef = firestore
                            .collection("users")
                            .document(userId!!)
                            .collection("Notes")

                        noteRef.document(noteId!!).delete()
                            .addOnSuccessListener {
                                Toast.makeText(this@diary_activity, "Note deleted successfully", Toast.LENGTH_SHORT).show()
                                // Optionally, finish the activity or update the UI
                                finish()
                            }
                            .addOnFailureListener { e ->
                                Toast.makeText(this@diary_activity, "Error deleting note: ${e.message}", Toast.LENGTH_SHORT).show()
                            }
                    }
                    .setNegativeButton("No", null)
                    .show()
            }
        }


        save_button.setOnClickListener{

            title=titleEditText.text.toString()
            content=noteContentEditText.text.toString()
            if(title.isNotEmpty()||content.isNotEmpty()){
                saveNote(title,content,noteId)
                Toast.makeText(this@diary_activity, "Note saved", Toast.LENGTH_SHORT).show()


            }else{
                Toast.makeText(this@diary_activity, "Empty Note title and content ", Toast.LENGTH_SHORT).show()

            }


        }
        //custom code goes here
        back_vector!!.setOnClickListener { finish() }
    }

    private fun loadNoteContent() {
        userId?.let { uid ->
            val noteRef = FirebaseFirestore.getInstance()
                .collection("users")
                .document(uid)
                .collection("Notes")
                .document(noteId!!)

            noteRef.get()
                .addOnSuccessListener { documentSnapshot ->
                    if (documentSnapshot.exists()) {
                        // Convert document to Note object
                        val note = documentSnapshot.toObject(Note::class.java)

                        // Populate UI with note data
                        note?.let {
                            titleEditText.setText(it.title)
                            noteContentEditText.setText(it.content)
                        }
                    } else {
                        Log.e("loadNoteContent", "No note found with ID: $noteId")
                    }
                }
                .addOnFailureListener { exception ->
                    Log.e("loadNoteContent", "Error loading note content: ${exception.message}", exception)
                }
        }


    }


fun saveNote(title: String, content: String, noteId: String?) {
    if (userId != null) {
        // Reference to Firestore
        val noteRef = firestore
            .collection("users")
            .document(userId!!)
            .collection("Notes")

        // Check if noteId is null to determine if it's a new note or an update
        if (noteId == null) {
            // Generate a unique document ID for the new note
            val newNoteId = noteRef.document().id

            val note = Note(
                title = title,
                timestamp = System.currentTimeMillis(),
                content = content
            )

            // Save the new note in Firestore under the generated newNoteId
            noteRef.document(newNoteId)
                .set(note)
                .addOnSuccessListener {
                    Log.d("SaveNote", "New note saved successfully with ID: $newNoteId")
                }
                .addOnFailureListener { e ->
                    Log.e("SaveNote", "Error saving new note", e)
                }
        } else {
            // Update the existing note
            val note = Note(
                title = title,
                timestamp = System.currentTimeMillis(),
                content = content
            )

            // Update the note in Firestore
            noteRef.document(noteId)
                .set(note) // or you could use .update() if you only want to update specific fields
                .addOnSuccessListener {
                    Log.d("SaveNote", "Note updated successfully with ID: $noteId")
                }
                .addOnFailureListener { e ->
                    Log.e("SaveNote", "Error updating note", e)
                }
        }
    } else {
        Log.e("SaveNote", "User ID is null; cannot save note.")
    }
}


}
